Issue 33 - Opt-in newsletter

kushankr opened the pull request:
- User can sign up for a newsletter on registration
- A confirmation mail is sent to user to confirm newsletter subscription

kushankr commented:
Hi @rupalkhilari,

I have completed the coding for #33 . Can you please review and test my changes? If you do not mind, I think that it would be better if you can also upload my changes to 8002 port test server. I can test my changes there as well.

A word about the coding:
you need to put the following code in the file named 'approval_frame/mailchimp_api_key.py'

```
import mailchimp


def get_mailchimp_api():
    return mailchimp.Mailchimp('API Key provided by Aaron')
```

This file is in '.gitignore' as it contains sensitive information.

Right now, I have coded it so that two confirmation mails are sent - one for account activation and one for newsletter subscription. 

For this, I had to extend the in-built registration view provided by django-registration. I believe that to merge both newsletter confirmation and account activation mails, I have to extend 'ActivationView' instead and use 'double_option=false' in 'subscribe()' function (called within extended ActivationView).

But as you will see in the code, sometimes 'subscribe()' returns an error and I am showing this to user. This is only possible if I extend the registration view and call 'subscribe()' in the extended registration class. So, if I have to accommodate error-showing, I have to call 'subscribe()' during registration instead of 'activation'. If i had put 'double_optin' to false, this would have meant that the user is subscribed to newsletter without clicking to confirm the link in account activation mail. And I have a checkbox to opt-in to newsletter subscription -> in my custom registration view, I can capture if this checkbox is checked. in the activationview, there is no way to get status of our checkbox. 

One more thing - If I would have used ActivationView, it was difficult to test locally. And I saw one thing in mailchimp documentation that 'double_option' is default set to true (this sends confirmation mail) -> and it had a warning that abusing this option can lead to suspension of mailchimp account.

All this will get clear when you review and test my code, Rupal. I thought that for now, I can code for 2 separate emails. This will ensure that changes are uploaded to beta application soon and our app can tap into pool of users for newsletter subscription.

I would appreciate your feedback on my PR. Thanks a lot, Rupal, for your time.

rupalkhilari commented:
Thanks Kushank for taking care of this feature and for the detailed explanation. I shall review it and get back to you by tomorrow.

kushankr commented:
Thanks, Rupal

rupalkhilari commented on .gitignore:
```diff
@@ -10,3 +10,4 @@ approval_frame/local_settings.py
 env
 Thumbs.db
 /staticfiles/
+approval_frame/mailchimp_api_key.py
```
It would be nice if you could keep the mailchimp_api_key module in the repository and have it read the mailchimp API key through a settings variable. Removing this module currently gives an ImportError while running the project, so that might be an issue when others try to run the project locally.

rupalkhilari commented on approval_frame/forms.py:
```diff
@@ -22,6 +22,26 @@ class RegistrationFormCustom(RegistrationFormUniqueEmail):
+        if newslettercheckbox:
+            if zipcode:
+                if len(zipcode) != 5:
+                    msg = "Please enter a zip code (5 digits)"
```
Unrelated, but it seems odd that the newsletter mandates a 5 character zip code, while countries like India and Canada use 6 characters. It would be nice to know from Aaron.

rupalkhilari commented on approval_frame/views.py:
```diff
@@ -36,3 +41,23 @@ def changeUsernameDone(request):
+    form_class = RegistrationFormCustom
+
+    def form_valid(self, request, form):
+        form = self.form_class(request.POST)
```
I seem to get this error after registering with a valid mailchip_api_key: 
```
  File "/usr/local/lib/python2.7/dist-packages/registration/views.py", line 82, in form_valid
  new_user = self.register(request, **form.cleaned_data)
 AttributeError: 'RegistrationFormCustom' object has no attribute 'cleaned_data'
```

Got to know from [this post](http://stackoverflow.com/questions/8090891/form-object-has-no-attribute-cleaned-data) that perhaps you need to check for form.is_valid() for this attribute to be made.

rupalkhilari commented:
Thanks @kushankr for implementing this feature. Apart from a couple of small issues which i have mentioned inline, it seems to be working well and looking good. 
Please let me know what you think. 
Thanks!

kushankr commented:
@rupalkhilari 

Thanks a lot for taking the time to review my changes. I will work on your comments.

Regarding the zipcode thing, you are right. I added this validation because I saw that while submitting the newsletter through browser (instead of app), it shows the validation error on invalid zip code input- "Please enter a zip code (5 digits)". And I just noticed that may be I should use 'IntegerField' instead of 'CharField'.

Initially, I thought that such restrictions should not be coded as after some research, I found out that a zip code (depending on country of the user) can contain digits or letter or even some special characters. Ideally, 'zipcode' should be a 'CharField' with a maximum length of 32 characters. However, I added this validation because I want to keep the format of zip code consistent across platforms.

I will contact Aaron regarding this. Btw, I think that the reason why this might not have come up earlier is because MailChimp API collects location data of subscribers by geocoding their IP address. So, if Aaron might have wanted to see the geographical distribution of newsletter subscribers, zipcode would not have figured. But anyways, I will contact him regarding this.

kushankr commented:
@rupalkhilari 

I have updated the PR. Can you please review and test it once? Thanks.

Regarding the settings comment, I have uploaded the module and updated README.md. To run this in development, you just need to put 'MAILCHIMP_API_KEY =  _Key provided by Aaron_ in local_settings.py. In production, this has to be declared in settings.py. 

Regarding the zipcode comment, I mailed Aaron (_cc_ ed you) and he told me that international residents can enter '00000' -> this is shown to user when the user enters a non 5 digit zip code.

Regarding the 'cleaned_data' attribute error comment -> unfortunately, I was unable to reproduce it. Your suggestion to fix it was very helpful. But I also learned that may be because I was re-instantiating the form, the new form did not have a 'cleaned_data' attribute. I have removed the line. I think that it should not show an error now. If it shows an error again, I can use 'form_valid()'.

Thank you, Rupal, for your feedback. I have tested my code and it seems to be working fine. Please let me know if you find any issues or you have any concerns. Thanks a lot.

rupalkhilari commented:
Thanks Kushank for addressing the comments. The changes work well. The issue with the registration form validation also appears to have been resolved with the fix. Thanks a lot !
